pkg/client/unversioned/request.go

```diff
@@ -839,7 +839,10 @@ func isTextResponse(resp *http.Response) bool {
 // checkWait returns true along with a number of seconds if the server instructed us to wait
 // before retrying.
 func checkWait(resp *http.Response) (int, bool) {
-	if resp.StatusCode != errors.StatusTooManyRequests {
+	switch r := resp.StatusCode; {
+	// any 500 error code and 429 can trigger a wait
+	case r == errors.StatusTooManyRequests, r >= 500:
```

Comment:
Retry-After is an HTTP header that indicates the server wants the client to
try again.  We widen the set of errors that we honor Retry-After (we don't
honor everything, for precisely that reason), but if the server gives us an
error and tells us to try again, we'll honor it.  We may need a client
behavior that says "i really, really, really never want to try again", but
the server is probably right.

On Thu, Nov 5, 2015 at 9:27 AM, David Eads <notifications@github.com> wrote:

> In pkg/client/unversioned/request.go
> <https://github.com/kubernetes/kubernetes/pull/16807#discussion_r44016403>
> :
>
> > @@ -839,7 +839,10 @@ func isTextResponse(resp *http.Response) bool {
> >  // checkWait returns true along with a number of seconds if the server instructed us to wait
> >  // before retrying.
> >  func checkWait(resp *http.Response) (int, bool) {
> > -	if resp.StatusCode != errors.StatusTooManyRequests {
> > +	switch r := resp.StatusCode; {
> > +	// any 500 error code and 429 can trigger a wait
> > +	case r == errors.StatusTooManyRequests, r >= 500:
>
> I liked what I saw until I got here. Why should our client retry on server
> errors? Can you tighten this to "500's that are also ServerTimeout errors"?
> That would require pulling the body.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16807/files#r44016403>.
>